common/utils/src/main/scala/org/apache/spark/internal/README.md

```diff
@@ -45,3 +45,29 @@ logger.error("Failed to abort the writer after failing to write map output.", e)
 ## Exceptions
 
 To ensure logs are compatible with Spark SQL and log analysis tools, avoid `Exception.printStackTrace()`. Use `logError`, `logWarning`, and `logInfo` methods from the `Logging` trait to log exceptions, maintaining structured and parsable logs.
+
+## External third-party ecosystem access
+
+* If you want to output logs in `scala code` through the structured log framework, you can define `custom LogKey` and use it in `scala` code as follows:
+
+```scala
+// External third-party ecosystem `custom LogKey` must be `extends LogKey`
+case object CUSTOM_LOG_KEY extends LogKey
+```
+```scala
+import org.apache.spark.internal.MDC;
+
+logInfo(log"${MDC(CUSTOM_LOG_KEY, "key")}")
+```
+
+* If you want to output logs in `java code` through the structured log framework, you can define `custom LogKey` and use it in `java` code as follows:
+
+```java
+// External third-party ecosystem `custom LogKey` must be `implements LogKey`
+public static class CUSTOM_LOG_KEY implements LogKey { }
+```
+```java
+import org.apache.spark.internal.MDC;
+
+logger.error("Unable to delete key {} for cache", MDC.of(CUSTOM_LOG_KEY, "key"));
+```
```

Comment:
Spark logging is not exposed api for users to directly leverage.

Reply:
If a third-party ecosystem wants to access a structured log framework, should it customize one?

Reply:
I am not sure I follow the question.

Is spark Logging being used today by third party ? Yes.
Do we support it ? No.
Do we disallow it ? Not really - power users put all sorts of things in org.apache.spark namespace and access internals, but we don't support that.

We are able to make these changes for structured logging, potentially breaking some usecases out there (even if we are making efforts to minimize impact) because this is an internal api.

Reply:
1.Let me take an example. In the production system, for example, an application named `SparkPi`, if you want to use `log4j` to print logs, there are two ways,
One is as follows(It does not need to be placed under the `org.apache.spark namespace`. `Logging` can also be used):
<img width="519" alt="image" src="https://github.com/apache/spark/assets/15246973/2f0c1cca-122c-4d3d-b67d-f64eae3f0ca1">

The other is:
<img width="497" alt="image" src="https://github.com/apache/spark/assets/15246973/04ba4b07-2442-4630-8aea-4e6e632bcaf4">

The last one is often used by `me`, and this is just `an application`.

If a third-party system, such as `iceberg`, wants to summarize logs through `a structured log framework`, it should be easy to do so by using the current structured framework.

2.In fact, our logging does not indicate that it is `any internal API` (We just didn't mark it as Developer Api), as shown below:
https://github.com/apache/spark/blob/fa8aa571ad18441622bb7e3ac66032ab9e7cbc0a/common/utils/src/main/scala/org/apache/spark/internal/Logging.scala#L84

I am worried that many users have used it like me, because it is direct and convenient (`conf/log4j.properties` can be used together)

Reply:
Like I said, are people using it ? Yes.
Is it exposed api, no.

The logging code is within 'internal' package.

Reply:
@mridulm the modified file here is `common/utils/src/main/scala/org/apache/spark/internal/README.md`, a README under the internal package.  It is not part of the Spark doc site.

I don't have a strong opinion on this though. If you insist we should NOT put this into the doc here, we should enhance the scala/java doc instead.

Reply:
Let us drop this doc - this feature is essentially private to spark. Let us avoid documenting it such that it gives an impression that users can directly leverage it.

Reply:
Well, I will remove the above `README.md` file and write the above guide to the `scala/java` doc.

Reply:
@mridulm @gengliangwang 
The `common/utils/src/main/scala/org/apache/spark/internal/README.md` file has been dropped, and the content of the original document has been distributed to `java/scala` docs.